api/python/gdh.py

```diff
@@ -0,0 +1,101 @@
+# Depends: requests pandas
+# You may already have these installed, otherwise install using pip:
+#     pip install requests pandas
+
+# Example invocations:
+# >>> from gdh import cases, qs
+# >>> c1 = cases(qs(country="New Zealand"), "API KEY HERE")
+# >>> c2 = cases(qs(country="New Zealand", admin1="Auckland"), "API KEY HERE")
+#
+# You can use the cases_cached() function to cache the cases locally.
+# This is useful for rerunning the script which will then use the
+# cached version.
+
+import os
+import io
+import json
+import hashlib
+import datetime
+
+import requests
+import pandas as pd
+
+GDH_URL_DEFAULT = "https://data.covid-19.global.health"
+GDH_URL = os.getenv("GDH_URL", GDH_URL_DEFAULT)
+downloadAsync = "/api/cases/downloadAsync"
+
+# sync with data-serving/data-service/src/util/search.ts
+FILTERS = {
+    "curator",
+    "gender",
+    "nationality",
+    "occupation",
+    "country",
+    "outcome",
+    "caseid",
+    "uploadid",
+    "sourceid",
+    "sourceurl",
+    "verificationstatus",
+    "admin1",
+    "admin2",
+    "admin3",
+    "variant",
+    "dateconfirmedafter",
+    "dateconfirmedbefore",
+}
+
+
+def qs(**kwargs):
+    if not kwargs:
+        raise ValueError("At least one filter has to be specified")
+    if unknown_filters := set(kwargs) - FILTERS:
+        raise ValueError(f"Unknown filters: {unknown_filters}")
+    kwargs = {
+        field: (f"'{value}'" if (isinstance(value, str) and " " in value) else value)
+        for field, value in kwargs.items()
+    }
+    for date_field in ["dateconfirmedafter", "dateconfirmedbefore"]:
+        # convert to yyyy-mm-dd if in datetime.date format
+        if date_field in kwargs and isinstance(
+            kwargs[date_field], datetime.date
+        ):
+            kwargs[date_field] = kwargs[date_field].isoformat()
+    return " ".join(f"{field}:{value}" for field, value in kwargs.items())
+
+
+def cases(query, apikey, server=GDH_URL):
+    res = requests.post(
+        f"{server}{downloadAsync}",
+        data=json.dumps({"format": "csv", "query": query}),
+        headers={"Content-Type": "application/json", "X-API-Key": apikey},
+    )
+    if res.status_code != 200:
+        raise ConnectionError(res.text)
+    if "signedUrl" in res.text:
+        # country-export returns a gzip compressed file
+        signedUrl = json.loads(res.text)["signedUrl"]
+        return pd.read_csv(signedUrl, compression="gzip")
+    else:
+        with io.StringIO(res.text) as buf:
+            return pd.read_csv(buf)
+
+
+def cases_cachefile(query, server=GDH_URL, folder="cache"):
+    to_hash = f"{query.strip()}|{server}"
+    sha256 = hashlib.sha256(to_hash.encode("utf-8")).hexdigest()
+    return f"{folder}/{sha256}.csv"
+
+
+# cases_cached returns a cached copy of cases if it exists,
+# otherwise saves to cache
+def cases_cached(query, apikey, server=GDH_URL, refresh=False, folder="cache"):
```

Comment:
In general I find functions named as verbs and objects named as nouns to be a better pattern to follow...i.e. a function named something like `get_cached_cases` will do just that.